components/lib/nearby/src/main/java/mozilla/components/lib/nearby/NearbyConnection.kt

```diff
@@ -0,0 +1,313 @@
+import android.util.Log
+import androidx.annotation.UiThread
+import androidx.appcompat.app.AlertDialog
+import com.google.android.gms.nearby.Nearby
+import com.google.android.gms.nearby.connection.AdvertisingOptions
+import com.google.android.gms.nearby.connection.ConnectionInfo
+import com.google.android.gms.nearby.connection.ConnectionLifecycleCallback
+import com.google.android.gms.nearby.connection.ConnectionResolution
+import com.google.android.gms.nearby.connection.ConnectionsClient
+import com.google.android.gms.nearby.connection.DiscoveredEndpointInfo
+import com.google.android.gms.nearby.connection.DiscoveryOptions
+import com.google.android.gms.nearby.connection.EndpointDiscoveryCallback
+import com.google.android.gms.nearby.connection.Payload
+import com.google.android.gms.nearby.connection.PayloadCallback
+import com.google.android.gms.nearby.connection.PayloadTransferUpdate
+import com.google.android.gms.nearby.connection.PayloadTransferUpdate.Status
+import com.google.android.gms.nearby.connection.Strategy
+import java.nio.charset.StandardCharsets.UTF_8
+
+/**
+ * A class that can be run on two devices to allow them to connect. This supports sending a single
+ * message at a time in each direction. This does not have internal synchronization. It should be
+ * called from the UI thread or externally synchronized.
+ *
+ * @constructor Constructs a new connection, which will call [NearbyConnectionListener.updateState]
+ *     with an argument of type [ConnectionState.Isolated]. No further action will be taken unless
+ *     other methods are called by the client.
+ * @param context context needed to initiate connection, used only at start
+ * @param name name shown by this device to other devices
+ * @param authenticate whether to authenticate the connection (true) or make it automatically (false)
+ * @param listener listener to be notified of changes of state and message transmission
+ *
+ */
+@UiThread
+class NearbyConnection(
+    private val context: Context,
+    private val name: String,
+    private val authenticate: Boolean,
+    private val listener: NearbyConnectionListener
+) {
+    // Compile-time constants
+    private val TAG = "NearbyConnection"
+    private val PACKAGE_NAME = "mozilla.components.feature.nearby.NearbyConnection"
+    private val STRATEGY = Strategy.P2P_STAR
+
+    /**
+     * The state of the connection. Changes in state are communicated through
+     * [NearbyConnectionListener.updateState].
+     *
+     * @property name: an English-language name for the state
+     */
+    public sealed class ConnectionState(val name: String) {
+        object Isolated : ConnectionState("isolated")
+        object Advertising : ConnectionState("advertising")
+        object Discovering : ConnectionState("discovering")
+        class Authenticating(
+            // sealed classes can't be inner, so we need to pass in the connection
+            private val nearbyConnection: NearbyConnection,
```

Comment:
Since a listener can receive the `updateState`, I wonder if we would leak the `nearbyConnection` during authentication, and I'm not sure the `private` is enough to hide it.

It seems we're passing it in so that we can continue, the result of callback might be something we can pass back to `NearbyConnection` for the caller to complete? This could be something that's on `NearbyConnection` instead of on the sealed class.

```kotlin
fun requestConnection( // ...
  requestResult: () -> Boolean) {
  If (requestResult) {
    connection.acceptConnection()
  } else {
    connection.rejectConnection()
  }
}
```

Also see the [`PermissionFeature`](https://mozac.org/api/mozilla.components.support.base.feature/-permissions-feature/) on how we tried to solve permission prompt requests to the user as an example API.

Reply:
Thank you for thinking about these issues. Two questions:

1. Why isn't making `nearbyConnection` private enough to hide it? Reflection?
2. Is there any need to protect the library from deliberate misuse? I don't see how misusing the interface would lead to security problems, just possible incorrect behavior on the local device.

I don't understand your proposed solution, but your answers to the above question might make it moot.

Reply:
> Why isn't making nearbyConnection private enough to hide it? Reflection?

I haven't tested your PR locally in another app, but I assumed you can still access the instance of `nearbyConnection` from the class instance. If that's not the case, maybe that should be fine.

> Is there any need to protect the library from deliberate misuse?

Misuse can be APIs that allow more permissive control than what should be allowed. I think it's safer to give an exact API without a backdoor to alter the state where we can't make the same guarantees. For example, with the nearbyConnection, a consumer might think they're doing something correct by re-authenticating every time you open the app screen, but that may cause issues instead, with your state machine or otherwise.

Reply:
The client does have a reference to the `nearbyConnection`, so the horse is already out of the barn. Do you think the library should be refactored so the client has only the current state? You can see a sample client in this repo: https://github.com/espertus/nearby-test

Reply:
Sorry, I think I confuseed NearbyConnection with the internal Nearby SDK, as long as we're not exposing that, I think it should be fine. Looking at your example, the new API is ace!